### PR TITLE
Aggregation  and window refactor

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

**Comment by sopel39** [comment]
Does this require benchmarks to run?

**Comment by dain** [comment]
> Does this require benchmarks to be run?

It would be great if you could do that.  It shouldn't affect anything, but you never know.

**Comment by findepi** [comment]
What APIs are changed by this PR?

**Comment by dain** [comment]
> What APIs are changed by this PR?

There are no SPI changes.  There are quite a few changes to the internal function APIs, especially around aggregations and window functions

**Comment by findepi** [comment]
In case the two conflict, this should come after an older PR also changing aggregation functions -- https://github.com/trinodb/trino/pull/8738

cc @martint @kasiafi

**Review by electrum — COMMENTED** [review]
Looks good through commit "Change AggregationFunctionMetadata to have multiple intermediate types"

**Comment by electrum** [comment]
Typo "AggergationMetadata" in commit title

**Comment by electrum** [comment]
Typo "aggergation" in commit title

**Comment by electrum** [comment]
Looks good through commit "Add deterministic to ResolvedFunction"

File: core/trino-main/src/main/java/io/trino/operator/window/RegularWindowPartition.java [hunk 1]
```diff
@@ -68,15 +70,15 @@ public RegularWindowPartition(
 
         // reset functions for new partition
         WindowIndex windowIndex = new PagesWindowIndex(pagesIndex, partitionStart, partitionEnd);
-        for (FramedWindowFunction framedWindowFunction : windowFunctions) {
-            framedWindowFunction.getFunction().reset(windowIndex);
+        for (WindowFunction framedWindowFunction : windowFunctions) {
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/operator/window/RegularWindowPartition.java:73]: Rename variable

File: core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java [hunk 2]
```diff
@@ -607,11 +608,17 @@ private static void generateAddIntermediateAsCombine(
...
             List<FieldDefinition> lambdaProviderFields,
-            MethodHandle combineFunction,
+            Optional<MethodHandle> combineFunction,
             CallSiteBinder callSiteBinder,
             boolean grouped)
     {
         MethodDefinition method = declareAddIntermediate(definition, grouped);
+        if (combineFunction.isEmpty()) {
+            method.getBody()
+                    .append(BytecodeExpressions.newInstance(UnsupportedOperationException.class, constantString("Aggregation is not decomposable")))
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java:612]: Use static import. Same for below

**Pull request merged**